src/main.rs

```diff
@@ -364,16 +365,57 @@ impl LanguageServer for Backend {
         params: GotoDefinitionParams,
     ) -> Result<Option<GotoDefinitionResponse>> {
         info!("ts_query_ls goto_definition: {params:?}");
-        let uri = params.text_document_position_params.text_document.uri;
-        let pos = Position {
-            line: params.text_document_position_params.position.line,
-            character: 0,
+        let uri = &params.text_document_position_params.text_document.uri;
+        let Some(tree) = self.ast_map.get(uri) else {
+            warn!("No AST built for URI: {uri:?}");
+            return Ok(None);
+        };
+        let Some(rope) = self.document_map.get(uri) else {
+            warn!("No document built for URI: {uri:?}");
+            return Ok(None);
         };
-        let range = Range::new(pos, pos);
-        Ok(Some(GotoDefinitionResponse::Scalar(Location {
-            uri,
-            range,
-        })))
+        let cur_pos = params.text_document_position_params.position;
+        let Some(current_node) =
+            get_current_capture_node(tree.root_node(), lsp_position_to_ts_point(cur_pos, &rope))
+        else {
+            return Ok(None);
+        };
+
+        let query = Query::new(&QUERY_LANGUAGE, "(capture) @cap").unwrap();
+        let mut cursor = QueryCursor::new();
+        let provider = TextProviderRope(&rope);
+
+        let mut parser = Parser::new();
+        parser
+            .set_language(&QUERY_LANGUAGE)
+            .expect("Error setting language for Query parser");
+
+        let defs = get_references(
+            &tree.root_node(),
+            &current_node,
+            &query,
+            &mut cursor,
+            &provider,
+            &rope,
+        )
+        .filter(|node| {
+            if let Some(p) = node.parent() {
+                p.kind() != "parameters"
+            } else {
+                true
+            }
+        })
+        .map(|node| ts_node_to_lsp_location(uri, &node, &rope))
+        .collect::<Vec<Location>>();
+
+        if defs.is_empty() {
+            let range = Range::new(cur_pos, cur_pos);
+            return Ok(Some(GotoDefinitionResponse::Scalar(Location {
+                uri: uri.clone(),
+                range,
+            })));
+        }
```

Comment:
Can this be deleted? I may be misunderstanding what this is doing but I think it is ok if we return an empty vector here, looks like in nvim at least (haven't tested other editors) the behavior is basically the same as if we kept this code block. Basically just keeps the cursor in the same spot (but slightly more intuitive result because it tells you no definitions were found)

Reply:
Yeah now that you mention it, removing that block makes more sense. Definitely a fan of adding fewer lines as well :smile: